### PR TITLE
fix `Wrong html processing in the overridden document.write function` (close #1311)

### LINKED ISSUE #1311: Wrong html processing in the overridden document.write function
Site: http://www.sponichi.co.jp/
Example:
```javascript
document.write('<script>');
document.write('</scr');
document.write('ipt>');
document.write('<div></div>');
```
Html without hammerhead:
```html
<script></script><div></div>
```
Html with hammerhead:
```html
<script>ipt><div></div></script>
```

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @miherlosev @churkin

### CODE REVIEW

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/node/document/writer.js [hunk 1]
```diff
@@ -190,14 +191,38 @@ export default class DocumentWriter {
         nativeMethods.insertBefore.call(elWithContent.parentNode, beginMarker, elWithContent);
     }
 
+    _getStartsWithClosingTagRegExp (tagName) {
+        tagName = tagName.toLowerCase();
+
+        if (!this.cachedRegExps[tagName]) {
+            const regExpStrParts = [tagName.charAt(tagName.length - 1), '?'];
```
Comment (miherlosev) [inline src/client/sandbox/node/document/writer.js]: Move code, creating the regexp by tagName, to the separte function `createStartsWithClosingTagRegExp`

File: src/client/sandbox/node/document/writer.js [hunk 2]
```diff
@@ -190,14 +191,38 @@ export default class DocumentWriter {
         nativeMethods.insertBefore.call(elWithContent.parentNode, beginMarker, elWithContent);
     }
 
+    _getStartsWithClosingTagRegExp (tagName) {
+        tagName = tagName.toLowerCase();
+
+        if (!this.cachedRegExps[tagName]) {
```
Comment (miherlosev) [inline src/client/sandbox/node/document/writer.js]: `this.cachedStartsWithClosingTagRegExps`

File: src/client/sandbox/node/document/writer.js [hunk 3]
```diff
@@ -190,14 +191,38 @@ export default class DocumentWriter {
...
         const elWithContent = endMarker;
 
         DocumentWriter._setUnclosedElementFlag(elWithContent);
 
-        elWithContent.textContent = elWithContent.textContent.replace(END_REMOVE_RE, '') + this.pending;
+        elWithContent.textContent = elWithContent.textContent.replace(END_REMOVE_RE, '');
         endMarker                 = nativeMethods.createElement.call(document, END_MARKER_TAG_NAME);
-        this.pending              = '';
+
+        if (this.pending && !this._getStartsWithClosingTagRegExp(elWithContent.tagName).test(this.pending)) {
```
Comment (miherlosev) [inline src/client/sandbox/node/document/writer.js]: It is difficult to read. Let's refactor as
```js
if (this.pending) {
    const startsWithClosingTagRegExp = this._getStartsWithClosingTagRegExp(elWithContent.tagName);
    .....
}
```

File: src/client/sandbox/node/document/index.js [hunk 4]
```diff
@@ -49,9 +49,14 @@ export default class DocumentSandbox extends SandboxBase {
     }
 
     attach (window, document) {
-        if (!this.documentWriter || this.window !== window || this.document !== document)
+        if (!this.documentWriter || this.window !== window || this.document !== document) {
```
[the remark this answers is not in the document]
Comment (LavrovArtem) [reply src/client/sandbox/node/document/index.js]: Tests fails in firefox and ie.
https://travis-ci.org/testcafe-build-bot/testcafe-hammerhead/jobs/276863389

File: src/client/sandbox/node/document/writer.js [hunk 5]
```diff
@@ -190,14 +192,45 @@ export default class DocumentWriter {
...
 
-        elWithContent.textContent = elWithContent.textContent.replace(END_REMOVE_RE, '') + this.pending;
+        elWithContent.textContent = elWithContent.textContent.replace(END_REMOVE_RE, '');
         endMarker                 = nativeMethods.createElement.call(document, END_MARKER_TAG_NAME);
-        this.pending              = '';
+
+        if (this.pending) {
+            const startsWithClosingTagRegExp = this._getStartsWithClosingTagRegExp(elWithContent.tagName);
+
+            if (!startsWithClosingTagRegExp.test(this.pending)) {
```
Comment (miherlosev) [inline src/client/sandbox/node/document/writer.js]: Difficult to read anyway.
I propose to use seprate variable
```js
const startsWithClosingTagRegExp                       = this._getStartsWithClosingTagRegExp(elWithContent.tagName);
const isPendingStartsWithClosingTagPartOfElWithContent = startsWithClosingTagRegExp.test(this.pending);
if (!isPendingStartsWithClosingTagPartOfElWithContent) {
  ...
}
```

**Comment by LavrovArtem** [comment]
ping @churkin

**Comment by LavrovArtem** [comment]
ping @churkin

**Pull request merged**